### PR TITLE
Compress pending messages

### PR DESCRIPTION
**Pull request by peterbryntesson**
Added the ability to compress pending alarm messages if so desired, new setting CompressedPayload in PendingAlarmsOptionsModel

### CODE REVIEW

File: api/src/Microsoft.Azure.IIoT.OpcUa.Api/src/Core/Models/PendingAlarmsOptionsApiModel.cs [hunk 1]
```diff
@@ -31,5 +31,11 @@ public class PendingAlarmsOptionsApiModel {
         [DataMember(Name = "snapshotInterval", Order = 2,
             EmitDefaultValue = false)]
         public int? SnapshotInterval { get; set; }
+
+        /// <summary>
+        /// Should we compress messages using GZip?
+        /// </summary>
+        [DataMember(Name = "compressedPayload", Order = 3)]
+        public bool CompressedPayload { get; set; } = false;
```
Comment (simonjaeger) [inline api/src/Microsoft.Azure.IIoT.OpcUa.Api/src/Core/Models/PendingAlarmsOptionsApiModel.cs:39]: Can we use the same format as the encoders ("UseReversibleEncoding", "UseAdvancedEncoding") for this... like "UseCompressedPayload" or "UseCompression"?

Comment (peterbryntesson) [reply api/src/Microsoft.Azure.IIoT.OpcUa.Api/src/Core/Models/PendingAlarmsOptionsApiModel.cs:39]: No, this was an explicit ask from Alexander to use the same as in 3.0.

**Review by simonjaeger — APPROVED** [review]

**Review by buzzfrog — APPROVED** [review]
LGTM

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs [hunk 2]
```diff
@@ -105,7 +106,16 @@ public class NetworkMessageEncoder : IMessageEncoder {
                     };
                     notification.Encode(helperEncoder);
                     helperEncoder.Close();
-                    var notificationSize = Encoding.UTF8.GetByteCount(helperWriter.ToString());
+                    var notificationSize = 0;
+                    if (compressedPayload) {
+                        var stream = new MemoryStream(Encoding.UTF8.GetBytes(helperWriter.ToString()));
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:111]: Please use using or try finally to ensure dispose

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs [hunk 3]
```diff
@@ -135,11 +145,19 @@ public class NetworkMessageEncoder : IMessageEncoder {
                         encoder.WriteEncodeable(null, element);
                     }
                     encoder.Close();
+
+                    var body = Encoding.UTF8.GetBytes(writer.ToString());
+                    if (compressedPayload) {
+                        var stream = new MemoryStream(body);
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:151]: Please use using or try finally to ensure dispose

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs [hunk 4]
```diff
@@ -135,11 +145,19 @@ public class NetworkMessageEncoder : IMessageEncoder {
...
+                        body = Encoding.UTF8.GetBytes(Convert.ToBase64String(stream.Zip()));
+                        stream.Dispose();
+                    }
+
                     var encoded = new NetworkMessageModel {
-                        Body = Encoding.UTF8.GetBytes(writer.ToString()),
+                        Body = body,
                         ContentEncoding = "utf-8",
```
Comment (koepalex) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:158]: Shouldn't ContentEncoding "gzip" when compressedPayload?

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:158]: Good catch, I'll fix

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs (hunk 3 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:151]: Fixed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs:111]: Yes, fixed

**Review by koepalex — APPROVED** [review]
:shipit:

**Pull request merged**